### PR TITLE
Add pg ilike search to posts title and body

### PR DESCRIPTION
**Pull request by todtb**
#3540

- Add pg ilike search to `posts.title` and `posts.body`
- Add search form to `/journal` (browse)
- Add post search results page which shows results, paginated

## Browse:
image

## Results:
image

image

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -499,6 +499,7 @@
       get :for_user
     end
   end
+  get "/journal_search", to: "posts#search"
```
Comment (kueda) [inline config/routes.rb:502]: Let's make this `/posts/search` instead.

**Review by kueda — CHANGES_REQUESTED** [review]
Many thanks, working for me, and sort of meets my original need, but not quite. To do that I'd like the following:

- [ ] Add the search boxes to posts/index (right-aligned in posts/_header would work for me)
- [ ] posts/search should take `parent_type` and `parent_id` params to scope search to a particular journal (doesn't have to be implemented that way, but I want journal-specific search so I can find particular posts in the iNat (site) journal, or my (user) journal)
- [ ] posts/search should show what journal you're searching

Sorry, another case of me writing an under-specified issue.

**Comment by todtb** [comment]
Ok, reworked this a bit so that it applies to all `parent_types` (`User`, `Project`, `Site`). Since I was doing this I took the opportunity to create a `HasJournal` concern for those models, thus helping clean up the controller flow a bit.

Also removed original "Browse" search as I see that wasn't really the primary goal here and searching ALL might not be that helpful.

N.B.
- Does change `Project`
`has_many :journal_posts, class_name: "Post", as: :parent`
to
`has_many :journal_posts, class_name: "Post", as: :parent, dependent: :destroy`
Though I'm assuming it should be this way and consistent with other models now, I highlight his as it is a change.

## Project (new)
image
## Project (old)
image
## Site
image
## User (self)
image
## User (other)
image

File: app/views/posts/search.html.haml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+- content_for :title do
+  = @title = "#{t(:find_posts_for)} #{@parent.journal_display_name}"
```
Comment (kueda) [inline app/views/posts/search.html.haml:2]: Translators will not like this b/c they can't control the position of the journal name, they won't be able to control for journals that begin with a vowel or consonant, and they won't be able to control for gender of the journal name. Here's an alternative that skirts two of these issues:

```ruby
= @title = t( :find_posts_for, vow_or_con: @parent.journal_display_name[0].downcase, journal_name: @parent.journal_display_name )
```

And then in en.yml:

```yaml
en:
  find_posts_for: "Find Posts for: %{journal_name}"
```

Gender could still be a problem, but there's no perfect way to deal with that with arbitrary journal names.

File: app/views/posts/_header.html.haml [hunk 3]
```diff
@@ -2,32 +2,59 @@
...
+  .row
+    .col-md-12
+      - title = t( :x_project_journal, project: strip_tags( @parent.title ), name: t( :journal ) )
+      %h1= title
+      #post_header
+        %div
+          - if @display_project.curated_by?(current_user)
+            = link_to t(:new_post), new_project_journal_post_path(@parent.journal_slug), id: "newpostbutton",
+                class: "default button"
+        = render partial: 'search_form', locals: { parent_type: @parent.class.name, parent_id: @parent.id }
```
Comment (kueda) [inline app/views/posts/_header.html.haml:20]: I'm not a fan of the extraneous whitespace this creates below the title when the viewer can't see the "New Post" button. Could we remove the search form from the header at put it at the top of the right column instead?

**Review by kueda — CHANGES_REQUESTED** [review]
Works great, and I like the concern. Just a few more changes and I think we'll be there.

File: app/views/posts/search.html.haml (hunk 2 above)
Comment (todtb) [reply app/views/posts/search.html.haml:2]: Excellent call, and a great lesson for thinking in i18n

File: app/views/posts/_header.html.haml (hunk 3 above)
Comment (todtb) [reply app/views/posts/_header.html.haml:20]: With the right column being the "archives" section? Something like:
## own
image

## other user
image

## project
image


## w/o archives:
image
### blog
image

Comment (kueda) [reply app/views/posts/_header.html.haml]: Exactly. Those screenshots look good to me.

**Review by kueda — APPROVED** [review]

**Comment by kueda** [comment]
Thanks again!

**Pull request merged**